backends/tc/backend.cpp

```diff
@@ -1018,72 +1052,80 @@ bool ConvertToBackendIR::hasExecuteMethod(const IR::Type_Extern *extn) {
-                tcPipeline->addExternDefinition(externDefinition);
+    auto ts = decl_type->to<IR::Type_SpecializedCanonical>();
+    if (decl_type->is<IR::Type_Extern>() || (ts && ts->baseType->is<IR::Type_Extern>())) {
+        const IR::Type_Extern *extn;
+        if (decl_type->is<IR::Type_Extern>()) {
+            extn = decl_type->to<IR::Type_Extern>();
+        } else {
+            extn = ts->baseType->to<IR::Type_Extern>();
+        }
+        auto eName = extn->name;
+        auto find = ControlStructPerExtern.find(eName);
+        if (find == ControlStructPerExtern.end()) {
+            return;
+        }
+        IR::TCExtern *externDefinition;
+        auto instance = new struct ExternInstance();
+        instance->instance_name = decl->toString();
+
+        auto constructorKeys = processExternConstructor(extn, decl, instance);
+
+        // Get Control Path information if specified for extern.
+        auto controlKeys = processExternControlPath(extn, decl, eName);
+
+        bool has_exec_method = hasExecuteMethod(extn);
+
+        /* If the extern info is already present, add new instance
+           Or else create new extern info.*/
+        auto iterator = externsInfo.find(eName);
+        if (iterator == externsInfo.end()) {
+            struct ExternBlock *eb = new struct ExternBlock();
+            if (eName == "DirectCounter") {
+                eb->externId = "0x1A000000"_cs;
+            } else if (eName == "Counter") {
+                eb->externId = "0x19000000"_cs;
+            } else if (eName == "Digest") {
+                eb->externId = "0x05000000"_cs;
+                instance->is_num_elements = true;
+                instance->num_elements = 0;
+            } else if (eName == "Meter") {
+                eb->externId = "0x1B000000"_cs;
+            } else if (eName == "DirectMeter") {
+                eb->externId = "0x1C000000"_cs;
             } else {
-                auto eb = externsInfo[eName];
-                externDefinition = ((IR::TCExtern *)tcPipeline->getExternDefinition(eName));
-                externDefinition->numinstances = ++eb->no_of_instances;
-                instance->instance_id = eb->no_of_instances;
-                eb->eInstance.push_back(instance);
-            }
-            IR::TCExternInstance *tcExternInstance =
-                new IR::TCExternInstance(instance->instance_id, instance->instance_name,
-                                         instance->is_num_elements, instance->num_elements);
-            if (controlKeys.size() != 0) {
-                tcExternInstance->addControlPathKeys(controlKeys);
-            }
-            if (constructorKeys.size() != 0) {
-                tcExternInstance->addConstructorKeys(constructorKeys);
+                externCount += 1;
+                std::stringstream value;
+                value << "0x" << std::hex << externCount;
+                eb->externId = value.str();
```

Comment:
the format of this externId will be 0x...ExternCount. where as for new one it will be something like "0x1C000000"

Are we initializing ExternCount with "0x1C000000"?

Reply:
For some externs, externIDs is fixed like Meter, DirectMeter (which is hardcoded on line above). This code will be executed for externs which doesn't have specific extern ID or for Custom externs. Here externCount is an integer value, which is converted to hex and appended with "0x" to generate extern_id in hexadecimal format.

Reply:
ok